Migrate CI to Docker

gp201 opened the pull request:
## Overview

<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes #18259
2. This PR does the following: This PR migrates all the test except the e2e, acceptance and lighthouse tests. The next steps are to address the lighthouse tests not running locally as a separate PR and a new PR that addresses the issues with the failing e2e and acceptance tests on GitHub Actions. 
3. (For bug-fixing PRs only) The original bug occurred because: [Explain what
   the cause of the bug was, and which PR introduced it]

## Essential Checklist

- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I don't have permissions to assign reviewers directly).
- [x] The linter/Karma presubmit checks pass on my local machine, and my PR follows the [coding style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide)).
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)

## Testing doc (for PRs with Beam jobs that modify production server data)

<!--
If this PR affects production server data, please follow
[these instructions](https://github.com/oppia/oppia/wiki/Testing-jobs-and-other-features-on-production#submitting-a-pr-with-a-new-job-or-feature-that-requires-third-party-api)
and link to the job request doc here.

Otherwise, please delete this section.
-->

- [ ] A testing doc has been written: ... (ADD LINK) ...
- [ ] _(To be confirmed by the server admin)_ All jobs in this PR have been verified on a live server, and the PR is safe to merge.

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. (For changes involving responsiveness or adjustment
of UI elements, this should be a video that gradually resizes the viewport from wide
to narrow and back again.) If this PR is for a developer-facing feature, provide
the videos/screenshots of developer-facing interface. Please also include
videos/screenshots of the developer tools browser console, so that we can be sure
that there are no console errors.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

gp201 committed:
Miscellaneous fixes for docker

gp201 committed:
Update Makefile and common.py

gp201 committed:
Fix Permission issues

gp201 committed:
Lint fixes

gp201 committed:
Lint fixes

gp201 committed:
Migrate CI to docker

gp201 committed:
Merge branch 'develop' into docker-ci-1

gp201 committed:
Lint fixes

gp201 committed:
Update Makefile to run typescript checks using the correct script

gp201 committed:
Remove unnecessary "needs: build" from e2e_lighthouse_performance_acceptance_tests.yml

gp201 committed:
Add third party size check to Makefile

gp201 committed:
Refactor import statement in run_lint_checks.py

gp201 committed:
Fix typescript strict tests

gp201 committed:
Merge branch 'develop' into docker-ci-1

gp201 committed:
Fixes

gp201 committed:
Add prettier action

gp201 committed:
Update e2e_lighthouse_performance_acceptance_tests.yml and pre_commit_hook.py

gp201 committed:
Add fail-fast option to e2e_test job

gp201 committed:
Fix acceptance test

gp201 committed:
Update Lighthouse test command in Makefile

gp201 committed:
Remove Make stop

gp201 committed:
Merge branch 'develop' of https://github.com/gp201/oppia into docker-ci-1

gp201 committed:
Add run_acceptance_tests.py to .coveragerc and run_acceptance_tests_test to run_backend_tests.py

gp201 committed:
Add typings for puppeteer acceptance tests

gp201 committed:
Fix backend test

gp201 committed:
Add fail-fast option to acceptance_test job

gp201 committed:
Fix acceptance tests

gp201 committed:
Merge branch 'develop' of https://github.com/gp201/oppia into docker-ci-1

gp201 committed:
Enable video recording for e2e tests and add debug logs

gp201 committed:
Refactor video recording condition in wdio.conf.js

gp201 committed:
Test e2e fix

gp201 committed:
Fix failing e2e test in about page footer

gp201 committed:
Add pause before uploading audio files

gp201 committed:
Update DEFAULT_WAIT_TIME_MSECS for mobile devices

gp201 committed:
Update DEFAULT_WAIT_TIME_MSECS_FOR_NEW_TAB value

gp201 committed:
Lint fix

StephenYu2018 commented on scripts/run_backend_tests.py:
```diff
@@ -530,6 +533,26 @@ def main(args: Optional[List[str]] = None) -> None:
+                    'scripts.script_import_test',
+                    'scripts.build_test',
+                    'scripts.run_acceptance_tests_test'
+                    ]
```
Remove a 4 spaces indent here.

StephenYu2018 commented on core/tests/webdriverio_desktop/voiceoverUploadFeatures.js:
```diff
@@ -75,6 +75,7 @@ describe('Voiceover upload features', function () {
 
   it('should upload an audio file', async function () {
     await explorationEditorTranslationTab.openUploadAudioModal();
+    await browser.pause(10000);
```
Are we supposed to wait exactly 10 seconds after opening the upload audio modal? Or is this statement here just to make the test pass? If it's the latter, then I think it would be better to open a separate issue to investigate why the test would fail without the statement, and reference that issue in a comment above this line. 

Ditto for line 101 of this file, and for lines 56 and 58 of the `./playVoiceovers.js` file.

gp201 replied:
This is a temporary addition. This statement is just for the tests to pass since it consistently fails.

This is because e2e tests run slower in GitHub Actions + Docker. 

Ditto for all `browser.pause(10000);` additions.

StephenYu2018 replied:
@gp201 Are we going to have these statements replaced after merging this PR given that these additions are temporary? If so, maybe create an issue page to track all the E2E test failures originating from CI migration to Docker.

StephenYu2018 reviewed (changes requested):
2 concerns before approval.

gp201 committed:
Merge branch 'develop' of https://github.com/gp201/oppia into docker-ci-1

gp201 committed:
Merge branch 'docker-ci-1' of https://github.com/gp201/oppia into docker-ci-1

gp201 committed:
Merge branch 'develop' of https://github.com/gp201/oppia into docker-ci-1

gp201 committed:
Refactor Makefile and update acceptance test configuration

gp201 committed:
Update release constants file path

gp201 committed:
merge branch 'develop' of https://github.com/gp201/oppia into docker-ci-1

gp201 committed:
Remove release_constants_docker.json and update OppiaFooterComponent

gp201 committed:
Lint fix

gp201 committed:
Increase timeout for loading message in waitFor.js

jayam04 committed:
added fix to google module installing without __init__.py files causing linter fails

jayam04 commented:
@gp201 PTAL at [comment](https://github.com/oppia/oppia/pull/19962#discussion_r1528498281), need your response before I can move forward on it.

jayam04 committed:
added types to python function

jayam04 committed:
Revert "added types to python function"

This reverts commit 680bcfa32216f10617369e9b259c49abca694d6d.

jayam04 committed:
Revert "Revert "added types to python function""

This reverts commit 2d2f0fb89968f46045c8fdf8790ba7b30ccc1041.

jayam04 committed:
Revert "added fix to google module installing without __init__.py files causing linter fails"

This reverts commit 45fd210151d8e2bbd84c834668ed32870aca4bd6.

jayam04 committed:
added google module fix for docker

jayam04 committed:
updated acceptance test in makefile to copy images to build

jayam04 commented:
@seanlip, @StephenYu2018 Could you review implementation of fixing google module for docker made in [commit here](https://github.com/oppia/oppia/pull/19962/commits/06eb01c080bfa8b4756f5ccbaab59b697e37db9d). I will add backend tests if approved.

seanlip commented on scripts/linters/run_lint_checks.py:
```diff
@@ -641,6 +640,9 @@ def main(args: Optional[List[str]] = None) -> None:
 
     if not feconf.OPPIA_IS_DOCKERIZED:
         install_third_party_libs.main()
+    else:
```
Can you switch this, i.e. if feconf.OPPIA_IS_DOCKERIZED, then fix_google_module, else run main()?

That way we avoid the "not".

seanlip replied:
Also it feels a bit random for this to be in the linter. Can we do something like: run install_third_party_libs.main() in all cases, and at the start of install_third_party_libs.main() do the fix if needed?

jayam04 replied:
done! used install_third_party_libs.main() for all cases.

seanlip reviewed (changes requested):
Thanks @jayam04, took a look, left a couple notes!

StephenYu2018 commented on scripts/install_third_party_libs.py:
```diff
@@ -167,6 +167,29 @@ def compile_protobuf_files(proto_files_paths: List[str]) -> None:
                 r'^import (\w*_pb2 as)', r'from proto_files import \1')
 
 
+def fix_google_module(google_module_path: str) -> None:
```
The current function name `fix_google_module` is vague in terms of what we're trying to fix. Are we trying to fix all issues that involve google modules? If not, then I suggest renaming the function that addresses the aforementioned concern. From what I've read in the docs and the linked NDB issue thread, a better name would be `make_google_module_importable_by_python`.

jayam04 replied:
done!

StephenYu2018 reviewed (changes requested):
One concern.

jayam04 committed:
updated install_third_party_libs

jayam04 committed:
Merge branch 'develop' into docker-ci-1

StephenYu2018 commented:
@jayam04 I believe there is still one prior concern of mine that hasn't been addressed yet. It is the only thing stopping me from approving this PR.

jayam04 committed:
updated make clean to remove test containers

jayam04 committed:
Merge branch 'develop' into docker-ci-1

gp201 committed:
Update comment for run_tests.prettier

StephenYu2018 reviewed (approved):
LGTM!

jayam04 commented:
@seanlip replied to review [here](https://github.com/oppia/oppia/pull/19962#discussion_r1568768646)

U8NWXD commented on .coveragerc:
```diff
@@ -9,13 +9,23 @@ omit =
     *scripts/linters/test_files/*
     *__init__.py
     */vm_deps/*
+    */scripts/build.py
+    */scripts/install_python_dev_dependencies.py
+    */scripts/install_python_prod_dependencies.py
+    */scripts/install_third_party_libs.py
+    */scripts/setup.py
+    */scripts/setup_gae.py
+    */scripts/run_acceptance_tests.py
```
I assume that we're excluding these from the coverage checks because they'll disappear when we migrate to Docker? Why is their exclusion necessary, given that we should already have full test coverage for these files? If it is needed, please add a comment to explain

gp201 replied:
Added a comment.

U8NWXD commented on scripts/linters/pylint_extensions.py:
```diff
@@ -47,7 +47,7 @@
 
 ALLOWED_PRAGMAS_FOR_INLINE_COMMENTS: Final = [
-    'pylint:', 'isort:', 'type: ignore', 'pragma:', 'https:'
+    'pylint:', 'isort:', 'type: ignore', 'pragma:', 'https:', 'docker'
```
```suggestion
    'pylint:', 'isort:', 'type: ignore', 'pragma:', 'https:', 'docker:'
```

to match the style of the other pragmas

U8NWXD commented on tsconfig-strict.json:
```diff
@@ -92,5 +92,13 @@
     "extensions",
     "typings"
+  ],
+  "exclude": [
+    "core/tests",
+    "extensions/classifiers/proto",
+    "**/*.spec.ts",
+    "**/*Spec.ts",
+    "**/*.ajs.ts",
+    "core/templates/tests/unit-test-utils.ajs.ts"
   ]
```
Why are we ignoring new files from the strict typescript checks? It doesn't seem to me that whether we enforce strict type checking on these files should depend on whether the checks are running in Docker or not

Also, there's already an exclude key above. If these changes are needed, can you make the changes there?

gp201 replied:
Removed the files.

U8NWXD commented on core/tests/test_utils.py:
```diff
@@ -315,7 +315,7 @@ def get_storage_model_classes() -> Iterator[Type[base_models.BaseModel]]:
 
 
-def generate_random_hexa_str() -> str:
+def generate_random_hexa_str() -> str: # docker: no cover
```
Why are we not covering this function? It seems easy to test

gp201 replied:
This is only used in the build_test.py script https://github.com/oppia/oppia/blob/bcb13ffee6a4e3a290f7476952a92d06938d8583/scripts/build_test.py#L243-L244

Which will be removed post-Docker migration.

U8NWXD commented on .github/workflows/backend_tests.yml:
```diff
@@ -72,23 +61,14 @@ jobs:
         if: startsWith(github.head_ref, 'update-changelog-for-release') == false
         run: pip install coverage
+      - name: Initialze Containers
+        run: |
```
Can we skip this for changelog PRs like the other steps in this job?

gp201 replied:
Fixed

U8NWXD replied:
I don't think it's fixed yet. You need to add `if: startsWith(github.head_ref, 'update-changelog-for-release') == false` to this step

gp201 replied:
Fixed. Forgot to push.

U8NWXD commented on .github/workflows/lighthouse_accessibility.yml:
```diff
@@ -24,47 +24,22 @@ jobs:
-          [[ $occurrences == "1" ]]
-          # Replace the target string with the replacement string.
-          sed -i 's/$SEARCH_TARGET/$REPLACEMENT/' $FILEPATH
-          echo "Replaced '$SEARCH_TARGET' with '$REPLACEMENT' in $FILEPATH"
```
Have you checked that this patch is no longer needed (e.g. by running the tests many times and getting 10+ consecutive passes)? We were seeing lighthouse test flakes before, which is why we added this patch

gp201 replied:
Added the patch.

seanlip reviewed (approved):
Thanks @jayam04, LGTM for the comment I left previously.

Re the tests:  you'll probably need to fix the backend and lint ones directly. For e2e tests, follow the usual rules: if it's a flake, check to see if any issue is open with that flake, and upvote it. If not and you're certain it's not due to your changes here, file an issue. In both cases, link to the issue(s) here and request a re-run if needed.

Do not use browser.pause() in e2e tests.

Feel free to ask if you have questions about the above.

Thanks!

seanlip commented:
@U8NWXD @Lawful2002 PTAL as codeowners, thanks.

U8NWXD commented on core/tests/puppeteer/lighthouse_setup.js:
```diff
@@ -460,7 +470,10 @@ const main = async function () {
-  const browser = await puppeteer.launch({headless: true});
+  const browser = await puppeteer.launch({
+    headless: true,
+    args: ['--no-sandbox'],
```
Please document why we need to disable the sandbox

U8NWXD commented on Makefile:
```diff
@@ -77,6 +77,7 @@ install_hooks:  ## Install required hooks
 	bash ./docker/pre_commit_hook.sh --install
 
 clean: ## Cleans the docker containers and volumes.
+	docker rm $$(docker ps -aq --filter="name=dev-server-run-") || true
```
This looks like it'll delete containers based on a name filter, which could be dangerous if developers have non-Oppia Docker containers on their systems. Why do we need this?

jayam04 replied:
@U8NWXD When running tests, containers with the name `dev-server-run-*` are created, which aren't deleted by docker compose command.

U8NWXD replied:
Okay, but there's still a problem because a user could create a container whose name follows the `dev-server-run-*` pattern. We don't want to delete containers that are user-created like this.

Instead, could we add a label (see https://docs.docker.com/reference/cli/docker/compose/run/) when issuing the `docker compose run ...` commands and then use that label to identify containers to delete? The label could be something like `oppia-generated-by-docker-compose-run` that's self-explanatory and unlikely to be chosen by a user for a manually-created container.

U8NWXD replied:
this still needs to be addressed

gp201 replied:
Fixed this to delete containers that uses the oppia-dev-server image

U8NWXD commented on Makefile:
```diff
@@ -122,19 +123,22 @@ logs.%: ## Shows the logs of the given docker service. Example: make logs.datast
 	docker compose run --no-deps --entrypoint "python -m scripts.third_party_size_check" dev-server
 
 run_tests.lint: ## Runs the linter tests
-	docker compose run --no-deps --entrypoint "/bin/sh -c 'git config --global --add safe.directory /app/oppia && python -m scripts.linters.run_lint_checks $(PYTHON_ARGS)'" dev-server || $(MAKE) stop
+	docker compose run --no-deps --entrypoint "/bin/sh -c 'git config --global --add safe.directory /app/oppia && python -m scripts.linters.run_lint_checks $(PYTHON_ARGS)'" dev-server
```
Here and elsewhere, why are we removing the finishing `make stop` command? Don't we want to shut down the containers once the tests complete?

gp201 replied:
I have removed `make stop` from these commands that used `docker compose` since the container gets terminated without `make stop`. I'll look into the remaining commands and update this thread.

U8NWXD commented on Makefile:
```diff
@@ -150,23 +154,24 @@ run_tests.check_overall_backend_test_coverage: ## Runs the check for overall bac
 run_tests.acceptance: ## Runs the acceptance tests for the parsed suite
 ## Flag for Acceptance tests
 ## suite: The suite to run the acceptance tests
+## MOBILE: Run acceptance test in mobile viewport.
```
```suggestion
## MOBILE: Whether to run the acceptance test in mobile viewport. Defaults to false.
```

U8NWXD commented on .github/workflows/e2e_lighthouse_performance_acceptance_tests.yml:
```diff
@@ -12,60 +12,10 @@ on:
-          path: /home/runner/work/oppia/build_files.zip
-          retention-days: 1
   e2e_test:
-    needs: build
```
Since we're not building ahead of time anymore, we have to re-run the webpack build for each test, right? What's the runtime impact of that?

gp201 replied:
The e2e tests are currently not running in prod mode under Docker. We will be enabling prod mode soon. This is currently being tracked here: https://github.com/oppia/oppia/issues/19096

U8NWXD commented on .github/workflows/e2e_lighthouse_performance_acceptance_tests.yml:
```diff
@@ -12,60 +12,10 @@ on:
-    needs: build
     runs-on:  ${{ matrix.os }}
     strategy:
+      fail-fast: false
```
Why turn this off? We enabled it to avoid saturating the CI queue with e2e jobs on PRs that had already failed CI checks. Ditto below

gp201 replied:
Fixed

U8NWXD commented on .github/workflows/e2e_lighthouse_performance_acceptance_tests.yml:
```diff
@@ -122,79 +72,32 @@ jobs:
-            xvfb-run -a --server-args="-screen 0, 1285x1000x24"
-            python -m scripts.run_e2e_tests --skip-install
-            --skip-build --suite=${{ matrix.suite.name }} --prod_env
+            sudo xvfb-run -a --server-args="-screen 0, 1285x1000x24" make run_tests.e2e suite=${{ matrix.suite.name }} VIDEO_RECORDING_IS_ENABLED=1
```
Unless you know that video recordings don't increase flakiness, please turn them off by default

gp201 replied:
Turned off

U8NWXD replied:
I still see `VIDEO_RECORDING_IS_ENABLED=1`, which seems to indicate that video recordings are still enabled

gp201 replied:
Fixed.

U8NWXD commented on .github/workflows/e2e_lighthouse_performance_acceptance_tests.yml:
```diff
@@ -207,20 +110,13 @@ jobs:
         with:
           name: webdriverio-screenshots
           path: /home/runner/work/oppia/webdriverio-screenshots
-      - name: Uploading webpack bundles as an artifact
-        if: ${{ failure() }}
-        uses: actions/upload-artifact@v3
-        with:
-          name: webpack-bundles
-          path: /home/runner/work/oppia/oppia/build
```
Can we keep this? It was added to help debug test failures

gp201 replied:
I updated the issue to have that information. When we make the change from dev mode to prod mode we will include this feature. https://github.com/oppia/oppia/issues/19096

U8NWXD commented on core/tests/puppeteer/lighthouse_setup.js:
```diff
@@ -505,6 +518,15 @@ const main = async function () {
   await generateDataForTopicAndStoryPlayer(browser, page);
   await generateDataForClassroom(browser, page);
+
+  fs.writeFileSync(
+    'core/tests/puppeteer/.env',
+    `exploration_id=${explorationId}\n` +
+      `story_id=${storyId}\n` +
+      `topic_id=${topicId}\n` +
+      `skill_id=${skillId}\n`
+  );
```
I think it's fine to transfer data this way instead of through the process's stdout like we used to do, but let's only use one method. Can you remove the code that this `.env` file makes redundant? For example: https://github.com/oppia/oppia/blob/03fc029568cc0c181bb9001f83db434c90087b67/scripts/run_lighthouse_tests.py#L109-L112

gp201 replied:
These lines will be removed with the PR for https://github.com/oppia/oppia/issues/18260
We're temporarily supporting both Python and Docker while devs transfer to the Docker setup. I updated the issue to include this info.

U8NWXD reviewed (changes requested):
Could you comment on the performance impact of these changes? How are the overall runtime on CI and CI runner load affected?

U8NWXD commented on core/tests/webdriverio_desktop/playVoiceovers.js:
```diff
@@ -53,7 +53,9 @@ describe('Voiceover player', function () {
     );
     await explorationEditorMainTab.setInteraction('EndExploration');
     await explorationEditorPage.navigateToTranslationTab();
+    await browser.pause(10000);
```
Please don't use browser.pause, as it's a common source of flakes

jayam04 replied:
done!

jayam04 committed:
Update Makefile

Co-authored-by: U8N WXD <U8NWXD@users.noreply.github.com>

jayam04 committed:
Update scripts/linters/pylint_extensions.py

Co-authored-by: U8N WXD <U8NWXD@users.noreply.github.com>

jayam04 committed:
removed browser.pause()

jayam04 commented:
@seanlip Created new issue for one e2e test failing. Could you review issue, do I need to add more info? Once done, I will open issues for other 5 failing tests.

seanlip commented:
Thanks @jayam04 -- the issue looks good, but I have some concerns about merging this with 6 known flakes that are likely to keep cropping up and blocking developers from merging things. It's going to make the dev workflow very tedious if developers have to keep rerunning the tests again and again.

I think it'd still be good to file the issues now, but after that, could you work with folks to try and fix them before we merge this PR? @jnvtnguyen has been doing some flake fixes recently, and might be a good person to ask for tips/help.

Thanks!

gp201 committed:
Update Docker CI configuration and remove unnecessary fail-fast option in backend_tests.yml and e2e_lighthouse_performance_acceptance_tests.yml

gp201 commented:
I have addressed some of the comments.
Will address the remaining by tomorrow:

1. https://github.com/oppia/oppia/pull/19962#discussion_r1577024885
2. https://github.com/oppia/oppia/pull/19962#discussion_r1579864271
3. https://github.com/oppia/oppia/pull/19962#discussion_r1579873250
4. https://github.com/oppia/oppia/pull/19962#discussion_r1579876645
5. https://github.com/oppia/oppia/pull/19962#discussion_r1579880172
6. https://github.com/oppia/oppia/pull/19962#discussion_r1579887478

gp201 committed:
chore: Patch Lighthouse source-maps.js to increase timeout for retrieving sourcemaps

gp201 committed:
chore: Add sudo to sed command in lighthouse_accessibility.yml workflow

gp201 committed:
chore: Remove redundant third party size check in Makefile

Lawful2002 reviewed (approved):
LGTM for codeowner files

gp201 commented:
Hi @U8NWXD,

Apologies for the trouble, and thanks for your patience. Since this PR is time-sensitive, could you quickly check out the other comments?

Right now, I'm focusing on resolving the comments provided in these two discussions:

1. https://github.com/oppia/oppia/pull/19962#discussion_r1577024885
2. https://github.com/oppia/oppia/pull/19962#discussion_r1579864271

U8NWXD commented:
I've replied to or resolved my responded-to comments

gp201 committed:
chore: Initialize Containers if head ref is not 'update-changelog-for-release'

gp201 committed:
chore: Disable video recording in e2e tests

jayam04 commented:
@gp201 Can you please re-run all e2e and acceptance tests.

gp201 committed:
Merge branch 'develop' into docker-ci-1

gp201 commented on .github/workflows/e2e_lighthouse_performance_acceptance_tests.yml:
```diff
@@ -12,7 +12,65 @@ on:
 
 jobs:
+  check_test_suites_to_run:
+    uses: ./.github/workflows/check_test_suites_to_run.yml
```
Update this to run in Docker

gp201 commented on .github/workflows/e2e_lighthouse_performance_acceptance_tests.yml:
```diff
@@ -12,7 +12,65 @@ on:
+          architecture: 'x64'
+          cache: 'pip'
+          cache-dependency-path: |
+            requirements.in
+            requirements.txt
+            requirements_dev.in
+            requirements_dev.txt
+      - uses: ./.github/actions/merge
+      - name: Cache node modules and third_party/static
+        uses: actions/cache@v3
+        env:
+          cache-name: cache-node-modules
+        with:
+          path: |
+            /home/runner/work/oppia/yarn_cache
+            /home/runner/work/oppia/oppia/third_party/static
+          key: ${{ runner.os }}-build-${{ env.cache-name }}-${{ hashFiles('yarn.lock', 'dependencies.json') }}
+          restore-keys: |
+            ${{ runner.os }}-build-${{ env.cache-name }}-
+            ${{ runner.os }}-build-
+            ${{ runner.os }}-
+      - uses: ./.github/actions/install-oppia-dependencies
+      - name: Install chrome
+        run: python -m scripts.install_chrome_for_ci
+      - name: Install ffmpeg
+        run: sudo apt install ffmpeg
+      - name: Build Webpack
+        run: python -m scripts.build --prod_env
+      - name: Zip build files
+        # We avoid using ../ or absolute paths because unzip treats these as
+        # security issues and will refuse to follow them.
+        run: |
+          zip -rqy build_files.zip oppia/third_party oppia_tools oppia/build oppia/webpack_bundles oppia/proto_files oppia/app.yaml oppia/assets/hashes.json oppia/proto_files oppia/extensions/classifiers/proto/* oppia/backend_prod_files oppia/dist
+        working-directory: /home/runner/work/oppia
+      - name: Upload build files artifact
+        uses: actions/upload-artifact@v3
+        with:
+          name: build_files
+          path: /home/runner/work/oppia/build_files.zip
+          retention-days: 1
```
Remove the build step for Docker.

gp201 committed:
Fix CI issues

gp201 committed:
Merge remote-tracking branch 'upstream/develop' into docker-ci-1

jayam04 committed:
added longer audio file to e2e test failing in docker

jayam04 committed:
Merge branch 'develop' into docker-ci-1

gp201 committed:
chore: Add fail-fast option to e2e_test job

jayam04 committed:
Update e2e_lighthouse_performance_acceptance_tests.yml

jayam04 committed:
Merge branch 'develop' into docker-ci-1

gp201 committed:
Revert E2E github action changes

gp201 committed:
Revert removal of install_chrome_for_ci

gp201 committed:
Add install_chrome_for_ci to backend test shards

gp201 committed:
Merge branch 'develop' into docker-ci-1

gp201 committed:
Merge branch 'develop' into docker-ci-2

gp201 committed:
bug: Add --no-sandbox flag in run_lighthouse_tests.py to allow running tests as root

gp201 committed:
Fix sandbox issue in lighthouse_setup.js and run_lighthouse_tests.py

gp201 committed:
Fix sandbox issue in lighthouse_setup.js

This commit fixes an issue in the lighthouse_setup.js file. The sandbox was causing problems when running tests as a root user in a Docker environment. To resolve this, the sandbox has been disabled by adding the  option to the  function.

gp201 committed:
Fix sandbox issue in run_lighthouse_tests.py and add --no-sandbox flag

gp201 committed:
Merge branch 'develop' into docker-ci-1

gp201 committed:
Increase default timeout

gp201 committed:
Revert changes

gp201 committed:
Merge branch 'develop' into docker-ci-1

gp201 committed:
Revert lighthouse tests on GitHub Actions

jayam04 committed:
Reverted to pre-commit hook, as it's already fixed in develop.

U8NWXD commented on scripts/run_lighthouse_tests.py:
```diff
@@ -196,7 +196,8 @@ def run_lighthouse_checks(lighthouse_mode: str) -> None:
         '--config=%s' % LIGHTHOUSE_CONFIG_FILENAMES[lighthouse_mode],
-        '--max-old-space-size=4096'
+        '--max-old-space-size=4096',
+        '--no-sandbox'
```
Why do we need to disable the sandbox? We should document the reasoning behind this decision

gp201 replied:
Removed. This will be added to the PR that fixes the lighthouse tests.

U8NWXD commented on .github/workflows/eslint_tests.yml:
```diff
@@ -26,26 +26,20 @@ jobs:
-            ${{ runner.os }}-
-      - name: Install Oppia dependencies
-        uses: ./.github/actions/install-oppia-dependencies
+      - name: Initialize Containers
```
Can we skip this for changelog PRs since we skip running the tests for those? ditto elsewhere

gp201 replied:
Fixed

U8NWXD commented on .github/workflows/backend_tests.yml:
```diff
@@ -27,31 +27,19 @@ jobs:
-        run: sudo apt-get install -y python-dev-is-python3 gcc
-      - name: Install coverage and configparser
+      - name: Free disk space
+        run: |
+          sudo rm -rf /usr/share/dotnet
+          sudo rm -rf "$AGENT_TOOLSDIRECTORY" 
+      - name: Initialze Containers
         if: startsWith(github.head_ref, 'update-changelog-for-release') == false
-        run: pip install coverage configparser
+        run: |
+          make build
+          sudo make run-devserver
+          make stop
```
Can we make these into actions that we reuse, like we did with `.github/actions/install-oppia-dependencies`? That would help reduce code duplication in the workflows

gp201 replied:
Done

gp201 committed:
Refactor GitHub Actions workflows to use a composite action for installing Oppia dependencies in Docker

gp201 committed:
Refactor Makefile to use more specific filter for removing Docker containers

gp201 committed:
Fix backend test

seanlip commented:
Adding @imchristie and @Vir-8 for codeowner review for oppia/acceptance-test-reviewers and oppia/lace-frontend-reviewers respectively.

imchristie reviewed (approved):
LGTM

Vir-8 reviewed (approved):
LGTM!
